tests/test_database_manager.py

```diff
@@ -0,0 +1,179 @@
+        await dbm_fx.update_row("testtableupdate", "AND", {"string1": "FizzBuzz"},
+                                {"string1": "thest"})
+
+        tmp = await dbm_fx.select_rows("testtableupdate", "AND", {"string1": "FizzBuzz"})
+        assert tmp[0][0] == "FizzBuzz"
+
+    async def test_delete_row(self, dbm_fx: DatabaseManager):
+        if not await dbm_fx.has_table("testtabledelete"):
+            await dbm_fx.create_table("testtabledelete", {"string1": "VARCHAR UNIQUE"})
+        await dbm_fx.insert_row("testtabledelete", {"string1": "thest"}, ("string1",))
+        await dbm_fx.delete_row("testtabledelete", "AND", {"string1": "thest"})
+        tmp = await dbm_fx.select_rows("testtabledelete", "AND", {"string1": "thest"})
+        assert len(tmp) == 0
+
+    async def test_double_dash_checks(self, dbm_fx: DatabaseManager):
+        with pytest.raises(ValueError):
+            await dbm_fx.delete_row("testtabledelete", "AND", {"test": "OR TRUE; -- "})
+            await dbm_fx.update_row("testtableupdate", "AND", {"string1": " OR FALSE; -- "})
+            await dbm_fx.insert_row("testtableinsert", {"string1": "stuff"}, ("string1",))
+            await dbm_fx.select_rows("testtableselect", "AND", {"test; -- ": "stuff"})
+
+        try:
+            await dbm_fx.delete_row("testtabledelete", "AND", {"test": "OR TRUE; -- "}, True)
+            await dbm_fx.update_row("testtableupdate", "AND", {"string1": " OR FALSE; -- "},
+                                    None, True)
+            await dbm_fx.insert_row("testtableinsert", {"string1": "stuff; -- "}, ("string1",), True)
+            await dbm_fx.select_rows("testtableselect", "AND", {"test; -- ": "stuff"}, True)
+        except pyodbc.Error:
+            pass
+
+    @pytest.mark.last
+    async def test_connection_error_handling(self, dbm_fx: DatabaseManager):
+        # disable the module, causing it to disconnect
+        dbm_fx.enabled = False
+
+        # this raises as we invoke it while it is disconnected
+        with pytest.raises(RuntimeError):
+            await dbm_fx.select_rows("testtableselect", "AND", {"string1": "thest"})
+
+        # now we reconnect
+        dbm_fx.enabled = True
+
+        # this should work now, as we reconnected
+        await dbm_fx.select_rows("testtableselect", "AND", {"string1": "thest"})
+
+        # now we bypass the shutdown code and close the connection()
+        dbm_fx.connection.close()
+
+        # this won't fail as it reconnects
+        await dbm_fx.select_rows("testtableselect", "AND", {"string1": "thest"})
+
+        # store the old cursor and connection
+        old_cursor = dbm_fx.cursor
+        old_conn = dbm_fx.connection
+
+        # this is a fake cursor
+        class I_am_a_duck():
+            def execute(self, *args, **kwargs):
+                raise pyodbc.Error("('08S01', '[08S01] SQLExecDirectW unable due to the connection"
+                                   " lost (35) (SQLExecDirectW)')")

```

Comment:
use a monkeypatch and an function definition here, rather than rolling a mock class

Reply:
We had a discussion about this, which started as we couldn't make this work with Monkeys iirc.

Reply:
I am certain a monkeypatch will work here.

```py
async def duck_execute(*args, **kwargs)
    raise pyodbc.Error("('08S01', '[08S01] SQLExecDirectW unable due to the connection"
                                   " lost (35) (SQLExecDirectW)')")

monkeypatch.setattr(dbm_fx.cursor, "execute", duck_execute)
```